### PR TITLE
Change ContainerAwareCommand to Command

### LINKED ISSUE #1270: TwoStepVerificationCommand using deprecated class
#### Sonata packages

```
$ composer show --latest 'sonata-project/*'
sonata-project/admin-bundle              3.78.1 3.78.1 The missing Symfony Admin Generator
sonata-project/block-bundle              3.21.0 3.21.0 Symfony SonataBlockBundle
sonata-project/cache                     2.0.1  2.0.1  Cache library
sonata-project/datagrid-bundle           3.2.0  3.2.0  Symfony SonataDatagridBundle
sonata-project/doctrine-extensions       1.10.1 1.10.1 Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.24.0 3.24.0 Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/exporter                  2.4.1  2.4.1  Lightweight Exporter library
sonata-project/form-extensions           1.6.0  1.6.0  Symfony form extensions
sonata-project/twig-extensions           1.4.1  1.4.1  Sonata twig extensions
sonata-project/user-bundle               4.9.0  4.9.0  Symfony SonataUserBundle

```

#### Symfony packages

```
$ composer show --latest 'symfony/*'
symfony/asset                      v4.4.15 v4.4.16 Symfony Asset Component
symfony/browser-kit                v4.4.15 v4.4.16 Symfony BrowserKit Component
symfony/cache                      v4.4.15 v4.4.16 Symfony Cache component with PSR-6, PSR-16, and tags
symfony/cache-contracts            v2.2.0  v2.2.0  Generic abstractions related to caching
symfony/config                     v4.4.15 v4.4.16 Symfony Config Component
symfony/console                    v4.4.15 v4.4.16 Symfony Console Component
symfony/css-selector               v4.4.15 v4.4.16 Symfony CssSelector Component
symfony/debug                      v4.4.16 v4.4.16 Symfony Debug Component
symfony/dependency-injection       v4.4.16 v4.4.16 Symfony DependencyInjection Component
symfony/doctrine-bridge            v4.4.15 v4.4.16 Symfony Doctrine Bridge
symfony/dom-crawler                v4.4.15 v4.4.16 Symfony DomCrawler Component
symfony/dotenv                     v4.4.15 v4.4.16 Registers environment variables from a .env file
symfony/error-handler              v4.4.16 v4.4.16 Symfony ErrorHandler Component
symfony/event-dispatcher           v4.4.16 v4.4.16 Symfony EventDispatcher Component
symfony/event-dispatcher-contracts v1.1.9  v2.2.0  Generic abstractions related to dispatching event
symfony/expression-language        v4.4.15 v4.4.16 Symfony ExpressionLanguage Component
symfony/filesystem                 v4.4.15 v4.4.16 Symfony Filesystem Component
symfony/finder                     v4.4.16 v4.4.16 Symfony Finder Component
symfony/flex                       v1.9.10 v1.9.10 Composer plugin for Symfony
symfony/form                       v4.4.15 v4.4.16 Symfony Form Component
symfony/framework-bundle           v4.4.15 v4.4.16 Symfony FrameworkBundle
symfony/http-client                v4.4.15 v4.4.16 Symfony HttpClient component
symfony/http-client-contracts      v2.3.1  v2.3.1  Generic abstractions related to HTTP clients
symfony/http-foundation            v4.4.16 v4.4.16 Symfony HttpFoundation Component
symfony/http-kernel                v4.4.16 v4.4.16 Symfony HttpKernel Component
symfony/inflector                  v4.4.15 v4.4.16 Symfony Inflector Component
symfony/intl                       v4.4.15 v4.4.16 A PHP replacement layer for the C intl extension that includes additional...
symfony/mailchimp-mailer           v4.4.15 v4.4.16 Symfony Mailchimp Mailer Bridge
symfony/mailer                     v4.4.15 v4.4.16 Symfony Mailer Component
symfony/maker-bundle               v1.22.0 v1.23.0 Symfony Maker helps you create empty commands, controllers, form classes,...
symfony/mime                       v4.4.16 v4.4.16 A library to manipulate MIME messages
symfony/monolog-bridge             v4.4.15 v4.4.16 Symfony Monolog Bridge
symfony/monolog-bundle             v3.6.0  v3.6.0  Symfony MonologBundle
symfony/options-resolver           v4.4.15 v4.4.16 Symfony OptionsResolver Component
symfony/orm-pack                   v1.2.0  v2.0.0  A pack for the Doctrine ORM
symfony/phpunit-bridge             v5.1.7  v5.1.8  Symfony PHPUnit Bridge
symfony/polyfill-ctype             v1.20.0 v1.20.0 Symfony polyfill for ctype functions
symfony/polyfill-intl-grapheme     v1.20.0 v1.20.0 Symfony polyfill for intl's grapheme_* functions
symfony/polyfill-intl-icu          v1.20.0 v1.20.0 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-intl-idn          v1.20.0 v1.20.0 Symfony polyfill for intl's idn_to_ascii and idn_to_utf8 functions
symfony/polyfill-intl-normalizer   v1.20.0 v1.20.0 Symfony polyfill for intl's Normalizer class and related functions
symfony/polyfill-mbstring          v1.20.0 v1.20.0 Symfony polyfill for the Mbstring extension
symfony/polyfill-php72             v1.20.0 v1.20.0 Symfony polyfill backporting some PHP 7.2+ features to lower PHP versions
symfony/polyfill-php73             v1.20.0 v1.20.0 Symfony polyfill backporting some PHP 7.3+ features to lower PHP versions
symfony/polyfill-php80             v1.20.0 v1.20.0 Symfony polyfill backporting some PHP 8.0+ features to lower PHP versions
symfony/process                    v4.4.15 v4.4.16 Symfony Process Component
symfony/profiler-pack              v1.0.5  v1.0.5  A pack for the Symfony web profiler
symfony/property-access            v4.4.15 v4.4.16 Symfony PropertyAccess Component
symfony/property-info              v4.4.15 v4.4.16 Symfony Property Info Component
symfony/routing                    v4.4.15 v4.4.16 Symfony Routing Component
symfony/security-acl               v3.1.0  v3.1.0  Symfony Security Component - ACL (Access Control List)
symfony/security-bundle            v4.4.15 v4.4.16 Symfony SecurityBundle
symfony/security-core              v4.4.15 v4.4.16 Symfony Security Component - Core Library
symfony/security-csrf              v4.4.15 v4.4.16 Symfony Security Component - CSRF Library
symfony/security-guard             v4.4.15 v4.4.16 Symfony Security Component - Guard
symfony/security-http              v4.4.15 v4.4.16 Symfony Security Component - HTTP Integration
symfony/serializer                 v4.4.15 v4.4.16 Symfony Serializer Component
symfony/serializer-pack            v1.0.4  v1.0.4  A pack for the Symfony serializer
symfony/service-contracts          v2.2.0  v2.2.0  Generic abstractions related to writing services
symfony/stopwatch                  v4.4.15 v4.4.16 Symfony Stopwatch Component
symfony/string                     v5.1.7  v5.1.8  Symfony String component
symfony/templating                 v4.4.15 v4.4.16 Symfony Templating Component
symfony/test-pack                  v1.0.7  v1.0.7  A pack for functional and end-to-end testing within a Symfony app
symfony/translation                v4.4.15 v4.4.16 Symfony Translation Component
symfony/translation-contracts      v2.3.0  v2.3.0  Generic abstractions related to translation
symfony/twig-bridge                v4.4.15 v4.4.16 Symfony Twig Bridge
symfony/twig-bundle                v4.4.15 v4.4.16 Symfony TwigBundle
symfony/validator                  v4.4.15 v4.4.16 Symfony Validator Component
symfony/var-dumper                 v4.4.16 v4.4.16 Symfony mechanism for exploring and dumping PHP variables
symfony/var-exporter               v4.4.15 v4.4.16 A blend of var_export() + serialize() to turn any serializable data struc...
symfony/web-link                   v4.4.15 v4.4.16 Symfony WebLink Component
symfony/web-profiler-bundle        v4.4.15 v4.4.16 Symfony WebProfilerBundle
symfony/yaml                       v4.4.15 v4.4.16 Symfony Yaml Component
```

#### PHP version

```
$ php -v
PHP 7.2.34 (cli) (built: Oct 13 2020 11:39:37) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.2.34, Copyright (c) 1999-2018, by Zend Technologies
```

The "Sonata\UserBundle\Command\TwoStepVerificationCommand" class extends "Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand" that is deprecated since Symfony 4.2, use {@see Command} instead.

### PR DESCRIPTION
**Pull request by eerison**
## Subject

I am targeting this branch, because it's deprecated since [Symfony version 4.2](https://symfony.com/blog/new-in-symfony-4-2-important-deprecations#deprecated-containerawarecommand) and there is BC break.

Closes #1270.

## Changelog

```markdown
### Changed
- Changed `ContainerAwareCommand` to `Command` in  `src/Command/TwoStepVerificationCommand.php`
```

### CODE REVIEW

**Review by phansys — COMMENTED** [review]
Before making this change (that could be considered a BC break), I think we could update some other aspects, like making the commands [lazy](https://symfony.com/doc/current/console/commands_as_services.html#console-command-service-lazy-loading) or allowing `execute()` method to return integer values.

**Comment by eerison** [comment]
> Before making this change (that could be considered a BC break), I think we could update some other aspects, like making the commands [lazy](https://symfony.com/doc/current/console/commands_as_services.html#console-command-service-lazy-loading) or allowing `execute()` method to return integer values.

Hey @phansys definettly it'll `BC break` because in this command is using `$this->getContainer()` and after change this dependicy and there isn't more access to all symfony container ....

the question is ....

can I remove this


https://github.com/sonata-project/SonataUserBundle/blob/28d48991fbc31ee5297dca8874e0e86933e8d938/src/Command/TwoStepVerificationCommand.php#L76-L98


and if `helper` or `userManager` is null, I throw an exception?

like

https://github.com/sonata-project/SonataUserBundle/blob/28d48991fbc31ee5297dca8874e0e86933e8d938/src/Command/TwoStepVerificationCommand.php#L76-L78

File: src/Command/TwoStepVerificationCommand.php [hunk 1]
```diff
@@ -71,8 +69,9 @@ public function configure(): void
     /**
      * {@inheritdoc}
      */
-    public function execute(InputInterface $input, OutputInterface $output): void
+    public function execute(InputInterface $input, OutputInterface $output): int
```
Comment (phansys) [inline src/Command/TwoStepVerificationCommand.php:72]: Changing the return type is a BC break for whose are extending this method. IMO, removing the return type we can respect the BC promise.
```suggestion
    public function execute(InputInterface $input, OutputInterface $output)
```

File: src/Command/TwoStepVerificationCommand.php [hunk 2]
```diff
@@ -16,17 +16,16 @@
...
 use Sonata\UserBundle\Model\UserInterface;
-use Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand;
+use Symfony\Component\Console\Command\Command;
 use Symfony\Component\Console\Input\InputArgument;
 use Symfony\Component\Console\Input\InputInterface;
 use Symfony\Component\Console\Input\InputOption;
 use Symfony\Component\Console\Output\OutputInterface;
 
-/**
- * NEXT_MAJOR: stop extending ContainerAwareCommand.
```
Comment (phansys) [inline src/Command/TwoStepVerificationCommand.php:26]: As the comment says, it was supposed to be changed in the next major.
BTW, I think we must mark this command as final, using the `@final` annotation until the next major.

**Review by phansys — CHANGES_REQUESTED** [review]

File: src/Command/TwoStepVerificationCommand.php (hunk 2 above)
Comment (eerison) [reply src/Command/TwoStepVerificationCommand.php:26]: any idea when the `next major` will be released? anyone started to work on that?

**Comment by eerison** [comment]
I'll close this PR because it'll be fixed on next major version

File: src/Command/TwoStepVerificationCommand.php (hunk 2 above)
Comment (eerison) [reply src/Command/TwoStepVerificationCommand.php:26]: Or can I do this MR to `master` ?

File: src/Command/TwoStepVerificationCommand.php (hunk 1 above)
Comment (eerison) [reply src/Command/TwoStepVerificationCommand.php:72]: "or allowing execute() method to return integer values"

but to allow `execute` return  `int`, I need to change this return type!

File: src/Command/TwoStepVerificationCommand.php (hunk 2 above)
Comment (VincentLanglet) [reply src/Command/TwoStepVerificationCommand.php:26]: Yes you can.

The next major will be released when every NEXT_MAJOR will be resolve + some others things, son any help is good to take.

We first try to release SonataAdmin 4.0

Comment (eerison) [reply src/Command/TwoStepVerificationCommand.php:26]: @phansys I'm closing thing thread could you see again, my new changes?

File: src/Command/TwoStepVerificationCommand.php [hunk 3]
```diff
@@ -16,32 +16,21 @@
...
+    protected static $defaultName = 'sonata:user:two-step-verification';
 
-    /**
-     * @var ?UserManagerInterface
-     */
+    private $helper;
     private $userManager;
 
-    /**
-     * NEXT_MAJOR: make $helper and $userManager mandatory (but still nullable).
```
Comment (eerison) [inline src/Command/TwoStepVerificationCommand.php:41]: to make $helper and $userManger mandatory I need to remove this

https://github.com/sonata-project/SonataUserBundle/blob/28d48991fbc31ee5297dca8874e0e86933e8d938/.php_cs.dist#L47

Comment (VincentLanglet) [reply src/Command/TwoStepVerificationCommand.php:41]: Not sure.

This enforce to write `?Foo $foo = null` instead of `Foo $foo = null` but you can still write `?Foo $foo`

Comment (eerison) [reply src/Command/TwoStepVerificationCommand.php:41]: Yeah That's true!

Thank you

File: src/Command/TwoStepVerificationCommand.php (hunk 1 above)
Comment (eerison) [reply src/Command/TwoStepVerificationCommand.php]: Sorry I see in your suggestion, that you're removing the type-hint.

**Comment by eerison** [comment]
@phansys could you check again please 😊

File: src/Command/TwoStepVerificationCommand.php [hunk 4]
```diff
@@ -16,34 +16,23 @@
...
-     * @var ?Helper
-     */
-    private $helper;
+    protected static $defaultName = 'sonata:user:two-step-verification';
 
-    /**
-     * @var ?UserManagerInterface
-     */
+    private $helper;
     private $userManager;
```
Comment (VincentLanglet) [inline src/Command/TwoStepVerificationCommand.php:30]: Why did you remove the phpdoc ? We should keep it.

File: src/Command/TwoStepVerificationCommand.php [hunk 5]
```diff
@@ -16,34 +16,23 @@
...
-    /**
-     * NEXT_MAJOR: make $helper and $userManager mandatory (but still nullable).
-     */
     public function __construct(
-        ?string $name,
-        ?Helper $helper = null,
-        ?UserManagerInterface $userManager = null
+        ?Helper $helper,
+        ?UserManagerInterface $userManager,
+        ?string $name = null
```
Comment (VincentLanglet) [inline src/Command/TwoStepVerificationCommand.php:35]: I don't think the comment asked to change the order of the param.

And IMHO, if we're throwing an exception:
```
if (null === $this->helper || null === $this->userManager) {
    throw new \RuntimeException('You need to inject Helper and UserManager.');
}
```
I don't understand why we should allow `null` here.

Comment (VincentLanglet) [reply src/Command/TwoStepVerificationCommand.php:35]: @greg0ire You made this commit: https://github.com/sonata-project/SonataUserBundle/commit/27de0075caa18dd40987733258b94cf7acb122cb

Maybe you can explain ?

Comment (eerison) [reply src/Command/TwoStepVerificationCommand.php:35]: - if we're accepting `?` then it can be `null`, But I need those classes to this code work, But if it's required for this code work then it shouldn't be null! :confused: 

- I changed the orders because optional parameters should be in the end of parameters list :confused: (But I guess if put `?$name`, Symfony will pass a null value to there ... then I don't need to put a default value ...., I'll test it)

Comment (VincentLanglet) [reply src/Command/TwoStepVerificationCommand.php:35]: > * if we're accepting `?` then it can be `null`, But I need those classes to this code work, But if it's required for this code work then it shouldn't be null! 😕

I agree.
Unless someone has an argument against I would use `Helper $helper` and `UserManagerInterface $userManager`.

File: src/Command/TwoStepVerificationCommand.php [hunk 6]
```diff
@@ -71,36 +59,15 @@ public function configure(): void
     /**
      * {@inheritdoc}
      */
-    public function execute(InputInterface $input, OutputInterface $output): void
+    public function execute(InputInterface $input, OutputInterface $output)
```
Comment (phansys) [inline src/Command/TwoStepVerificationCommand.php:62]: Since you're currently targeting `master`, you can add the proper return type declaration.
```suggestion
    public function execute(InputInterface $input, OutputInterface $output): int
```

File: src/Command/TwoStepVerificationCommand.php [hunk 7]
```diff
@@ -16,34 +16,23 @@
...
 use Symfony\Component\Console\Input\InputArgument;
 use Symfony\Component\Console\Input\InputInterface;
 use Symfony\Component\Console\Input\InputOption;
 use Symfony\Component\Console\Output\OutputInterface;
 
-/**
- * NEXT_MAJOR: stop extending ContainerAwareCommand.
- */
-class TwoStepVerificationCommand extends ContainerAwareCommand
+class TwoStepVerificationCommand extends Command
```
Comment (phansys) [inline src/Command/TwoStepVerificationCommand.php:25]:
```suggestion
final class TwoStepVerificationCommand extends Command
```

**Review by phansys — CHANGES_REQUESTED** [review]

File: src/Command/TwoStepVerificationCommand.php (hunk 5 above)
Comment (greg0ire) [reply src/Command/TwoStepVerificationCommand.php:35]: > Maybe you can explain ?

@VincentLanglet  the end goal of 27de007 is to push people to use DI. `null` should not be allowed.

Comment (greg0ire) [inline src/Command/TwoStepVerificationCommand.php:35]:
```suggestion
        Helper $helper,
        UserManagerInterface $userManager,
        ?string $name = null
```

File: src/Command/TwoStepVerificationCommand.php (hunk 4 above)
Comment (eerison) [reply src/Command/TwoStepVerificationCommand.php]: Added again :) 

Do we need the docs to help with IDE?

File: src/Command/TwoStepVerificationCommand.php (hunk 7 above)
Comment (eerison) [reply src/Command/TwoStepVerificationCommand.php:25]: why this class can't be extended?

Comment (eerison) [reply src/Command/TwoStepVerificationCommand.php]: @phansys I've applied your suggestion, But could you explain why add "final" here?

Comment (VincentLanglet) [reply src/Command/TwoStepVerificationCommand.php]: For next major, we're trying to close the API by adding as many `final` keyword as possible.
This will allow us easier change to our base code.

Comment (eerison) [reply src/Command/TwoStepVerificationCommand.php]: ohhhh thank you for explain :smiley:

**Review by VincentLanglet — APPROVED** [review]

**Review by phansys — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Thanks @eerison
